File: client/src/main/scala/stasis/client/ops/recovery/stages/EntityProcessing.scala
package stasis.client.ops.recovery.stages

import java.nio.file.FileSystem

import scala.concurrent.ExecutionContext
import scala.concurrent.Future
import scala.util.Try
import scala.util.control.NonFatal
import scala.util.matching.Regex

import org.apache.pekko.NotUsed
import org.apache.pekko.stream.Materializer
import org.apache.pekko.stream.SharedKillSwitch
import org.apache.pekko.stream.scaladsl.Flow
import org.apache.pekko.stream.scaladsl.Source
import org.apache.pekko.util.ByteString

import stasis.client.encryption.secrets.DeviceSecret
import stasis.client.model.EntityMetadata
import stasis.client.model.TargetEntity
import stasis.client.model.TargetEntity.Destination
import stasis.client.ops.Metrics
import stasis.client.ops.ParallelismConfig
import stasis.client.ops.exceptions.EntityProcessingFailure
import stasis.client.ops.exceptions.OperationStopped
import stasis.client.ops.recovery.Providers
import stasis.client.ops.recovery.RecoveryEntityKind
import stasis.client.utils.StringPaths.StringAsPath
import stasis.core.packaging.Crate
import stasis.core.routing.exceptions.PullFailure
import stasis.shared.ops.Operation

trait EntityProcessing {
  protected def deviceSecret: DeviceSecret
  protected def providers: Providers
  protected def parallelism: ParallelismConfig

  protected implicit def mat: Materializer
  protected implicit def ec: ExecutionContext

  private val metrics = providers.telemetry.metrics[Metrics.RecoveryOperation]

  def entityProcessing(implicit
    operation: Operation.Id,
    killSwitch: SharedKillSwitch
  ): Flow[TargetEntity, TargetEntity, NotUsed] =
    Flow[TargetEntity]
      .filter {
        case TargetEntity(_, Destination.Default, _, _)                                       => true
        case TargetEntity(_, Destination.Directory(_, true), _, _)                            => true
        case TargetEntity(_, Destination.Directory(_, false), _: EntityMetadata.File, _)      => true
        case TargetEntity(_, Destination.Directory(_, false), _: EntityMetadata.Library, _)   => true
        case TargetEntity(_, Destination.Directory(_, false), _: EntityMetadata.Directory, _) => false
      }
      .map { entity =>
        RecoveryEntityKind.prepare(kinds = providers.kinds, entity = entity, providers = providers)
        entity
      }
      .mapAsync(parallelism.entities) {
        case entity if entity.hasContentChanged => processContentChanged(entity)
        case entity                             => processMetadataChanged(entity)
      }
      .wireTap { targetEntity =>
        metrics.recordEntityProcessed(entity = targetEntity)
        providers.track.entityProcessed(entity = targetEntity.destinationRef)
      }

  private def processContentChanged(
    entity: TargetEntity
  )(implicit killSwitch: SharedKillSwitch, operation: Operation.Id): Future[TargetEntity] =
    EntityProcessing.expectContentMetadata(entity).flatMap { contentMetadata =>
      val crates = pull(contentMetadata.crates, entity)
      implicit val prv: Providers = providers

      val decompressor = providers.compression.decoderFor(entity)

      providers.track.entityProcessingStarted(
        entity = entity.ref,
        expectedParts = crates.size
      )

      def recordPartProcessed(): Unit =
        providers.track.entityPartProcessed(entity = entity.ref)

      val content: Source[ByteString, NotUsed] =
        crates
          .decrypt(withPartSecret = deviceSecret.toFileSecret(_, contentMetadata.checksum))
          .merge(onPartProcessed = recordPartProcessed)
          .via(killSwitch.flow)
          .decompress(decompressor = decompressor)
          .wireTap(bytes =>
            metrics.recordEntityChunkProcessed(step = "decompressed", extra = decompressor.name, bytes = bytes.length)
          )

      RecoveryEntityKind
        .write(kinds = providers.kinds, entity = entity, content = content, providers = providers)
        .map(_ => entity)
        .recoverWith {
          case OperationStopped(e) => Future.failed(e)
          case NonFatal(e)         => Future.failed(EntityProcessingFailure(entity = entity.ref, cause = e))
        }
    }

  private def processMetadataChanged(entity: TargetEntity)(implicit operation: Operation.Id): Future[TargetEntity] = {
    providers.track.entityProcessingStarted(entity = entity.ref, expectedParts = 0)
    Future.successful(entity)
  }

  private def pull(crates: Map[String, Crate.Id], entity: TargetEntity): Iterable[(Int, String, Source[ByteString, NotUsed])] = {
    val sources = crates.map { case (partPath, crate) =>
      val partId = EntityProcessing.partIdFromPath(path = partPath, fs = providers.filesystem)

      val source: Source[ByteString, NotUsed] = Source
        .lazyFutureSource { () =>
          providers.clients.core
            .pull(crate)
            .flatMap {
              case Some(source) =>
                Future.successful(source)

              case None =>
                Future.failed(
                  PullFailure(
                    s"Failed to pull crate [${crate.toString}] for entity [${entity.originalRef.key}]"
                  )
                )
            }
        }
        .mapMaterializedValue(_ => NotUsed)

      (partId, partPath, source)
    }

    val lastPartId = sources.map(_._1).maxOption.getOrElse(0)
    require(
      lastPartId + 1 == crates.size,
      s"Unexpected last part ID [${lastPartId.toString}] encountered for an entity with [${crates.size.toString}] crate(s)"
    )

    sources
  }

  private type CratesSources = Iterable[(Int, String, Source[ByteString, NotUsed])]
  private type EntitySource = Source[ByteString, NotUsed]

  private implicit class DecryptedCrates(crates: CratesSources) extends internal.DecryptedCrates(crates)
  private implicit class MergedCrates(crates: CratesSources) extends internal.MergedCrates(crates)
  private implicit class DecompressedByteStringSource(source: EntitySource) extends internal.DecompressedByteStringSource(source)
}

object EntityProcessing {
  private val pathPartId: Regex = ".*__part=(\\d+)".r

  def partIdFromPath(path: String, fs: FileSystem): Int =
    path.extractName(fs) match {
      case pathPartId(id) => Try(Integer.parseInt(id)).getOrElse(0)
      case _              => 0
    }

  def expectContentMetadata(entity: TargetEntity): Future[EntityMetadata.WithContent] =
    entity.existingMetadata match {
      case content: EntityMetadata.WithContent =>
        Future.successful(content)

      case directory: EntityMetadata.Directory =>
        Future.failed(
          new IllegalArgumentException(
            s"Expected metadata for file but directory metadata for [${directory.path}] provided"
          )
        )
    }

}

File: client-android/lib/src/main/kotlin/stasis/client_android/lib/ops/recovery/stages/EntityProcessing.kt
package stasis.client_android.lib.ops.recovery.stages

import kotlinx.coroutines.flow.Flow
import kotlinx.coroutines.flow.filter
import kotlinx.coroutines.flow.map
import kotlinx.coroutines.flow.mapNotNull
import okio.Source
import stasis.client_android.lib.api.clients.exceptions.EndpointFailure
import stasis.client_android.lib.encryption.secrets.DeviceSecret
import stasis.client_android.lib.model.EntityMetadata
import stasis.client_android.lib.model.TargetEntity
import stasis.client_android.lib.model.core.CrateId
import stasis.client_android.lib.ops.OperationId
import stasis.client_android.lib.ops.recovery.Providers
import stasis.client_android.lib.ops.recovery.RecoveryEntityKind
import stasis.client_android.lib.ops.recovery.stages.internal.DecompressedSource.decompress
import stasis.client_android.lib.ops.recovery.stages.internal.DecryptedCrates.decrypt
import stasis.client_android.lib.ops.recovery.stages.internal.MergedCrates.merged
import stasis.client_android.lib.utils.NonFatal.nonFatal
import stasis.client_android.lib.utils.StringPaths.extractName
import stasis.client_android.lib.utils.Try
import java.nio.file.FileSystem
import java.nio.file.FileSystems

interface EntityProcessing {
    val deviceSecret: DeviceSecret
    val providers: Providers

    fun entityProcessing(operation: OperationId, flow: Flow<TargetEntity>): Flow<TargetEntity> =
        flow.filter { entity ->
            when (val destination = entity.destination) {
                is TargetEntity.Destination.Default -> true
                is TargetEntity.Destination.Directory -> when {
                    destination.keepDefaultStructure -> true
                    entity.existingMetadata is EntityMetadata.File -> true
                    entity.existingMetadata is EntityMetadata.Library -> true
                    else -> false
                }
            }
        }
            .map { entity ->
                RecoveryEntityKind.prepare(providers.kinds, entity, providers)
                entity
            }
            .map { entity ->
                entity to Try {
                    when {
                        entity.hasContentChanged -> processContentChanged(operation, entity)
                        else -> processMetadataChanged(operation, entity)
                    }
                }
            }
            .mapNotNull { (entity, result) ->
                when (result) {
                    is Try.Success -> {
                        providers.track.entityProcessed(
                            operation = operation,
                            entity = entity.destinationRef
                        )
                        entity
                    }

                    is Try.Failure -> {
                        providers.track.failureEncountered(
                            operation = operation,
                            entity = entity.ref,
                            failure = result.exception.nonFatal()
                        )
                        providers.analytics.recordFailure(result.exception)

                        if (result.exception is EndpointFailure) {
                            throw result.exception
                        } else {
                            null
                        }
                    }
                }
            }

    private suspend fun processContentChanged(operation: OperationId, entity: TargetEntity): TargetEntity {
        val contentMetadata = expectContentMetadata(entity)
        val crates = contentMetadata.crates

        providers.track.entityProcessingStarted(
            operation = operation,
            entity = entity.ref,
            expectedParts = crates.size
        )

        fun recordPartProcessed(): Unit =
            providers.track.entityPartProcessed(operation, entity = entity.ref)

        val content = pull(crates, entity.originalRef.key)
            .decrypt(withPartSecret = { deviceSecret.toFileSecret(it, contentMetadata.checksum) }, providers = providers)
            .merged(onPartProcessed = ::recordPartProcessed)
            .decompress(decompressor = providers.compression.decoderFor(entity))

        RecoveryEntityKind.write(providers.kinds, entity, content, providers)

        return entity
    }

    private fun processMetadataChanged(operation: OperationId, entity: TargetEntity): TargetEntity {
        providers.track.entityProcessingStarted(operation = operation, entity = entity.ref, expectedParts = 0)
        return entity
    }

    private suspend fun pull(
        crates: Map<String, CrateId>,
        entityKey: String
    ): List<Triple<Int, String, suspend () -> Source>> {
        val sources = crates.map { (partPath, crate) ->
            val source = suspend {
                when (val source = providers.clients.core.pull(crate)) {
                    null -> throw RuntimeException("Failed to pull crate [$crate] for entity [$entityKey]")
                    else -> source
                }
            }

            Triple(partIdFromPath(partPath, FileSystems.getDefault()), partPath, source)
        }

        val lastPartId = sources.maxOfOrNull { it.first } ?: 0
        require(lastPartId + 1 == crates.size) {
            "Unexpected last part ID [$lastPartId] encountered for an entity with [${crates.size}] crate(s)"
        }

        return sources
    }

    companion object {
        private val pathPartId: Regex = ".*__part=(\\d+)".toRegex()

        fun partIdFromPath(path: String, fs: FileSystem): Int =
            pathPartId.find(path.extractName(fs))?.groups?.get(1)?.value?.toIntOrNull() ?: 0

        fun expectContentMetadata(entity: TargetEntity): EntityMetadata.WithContent =
            when (val metadata = entity.existingMetadata) {
                is EntityMetadata.File -> metadata
                is EntityMetadata.Library -> metadata
                is EntityMetadata.Directory -> throw IllegalArgumentException(
                    "Expected metadata for file but directory metadata for [${metadata.path}] provided"
                )
            }
    }
}
